File: 2017/01/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};

fn get_count(numbers: &Vec<u32>, index_offset: usize) -> u32 {
    (0..numbers.len()).fold(0, |count, index| 
        if numbers[index] == numbers[(index + index_offset) % numbers.len()] { 
            count + numbers[index] 
        } else { 
            count 
        })
}

fn solve(numbers: &Vec<u32>) -> (u32, u32) {
    (get_count(numbers, numbers.len() - 1), get_count(numbers, numbers.len() / 2))
}

fn get_input(file_path: &String) -> Vec<u32> {
    let contents = fs::read_to_string(file_path).expect("Error reading input file!");
    contents.trim().chars().map(|c| c.to_string().parse::<u32>().unwrap()).collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P2: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2017/01/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef struct
{
    int *numbers;
    int size;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int getCount(Input input, int indexOffset)
{
    int count = 0;
    for (int index = 0; index < input.size; index++)
        if (input.numbers[index] == input.numbers[(index + indexOffset) % input.size])
            count += input.numbers[index];
    return count;
}

Results solve(Input input)
{
    return (Results){
        getCount(input, input.size - 1),
        getCount(input, input.size / 2)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    int length = ftell(file);
    rewind(file);
    int *numbers = calloc(length, sizeof(int));
    int *current = numbers;
    int c;
    while ((c = getc(file)) != EOF)
        if (c >= '0' && c <= '9')
            *current++ = c - '0';
        else
            length--;
    fclose(file);
    return (Input){numbers, length};
}

void freeInput(Input input)
{
    free(input.numbers);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}